### PR TITLE
Feature: add osu-ruleset pp scoring

### PR DESCRIPTION
**Pull request by JustRoxy**
Adds updatable and blazingly fast :rocket: :zap: pp processor for osu!Cringe :sunglasses:

### CODE REVIEW

File: Cringe.Bancho/Cringe.Bancho.csproj [hunk 1]
```diff
@@ -1,7 +1,8 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk.Web">
 
     <PropertyGroup>
-        <TargetFramework>net6.0</TargetFramework>
+        <TargetFramework>net8.0</TargetFramework>
```
Comment (stanriders) [inline Cringe.Bancho/Cringe.Bancho.csproj:4]: Please update `Dockerfile` as well

File: Cringe/Cringe.csproj [hunk 2]
```diff
@@ -2,25 +2,25 @@
...
-    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="6.0.35" />
-    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="6.0.35">
+    <PackageReference Include="AutoMapper" Version="13.0.1" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="9.0.0" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="9.0.0">
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
     <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
     <PackageReference Include="OppaiSharp" Version="3.4.0" />
```
Comment (stanriders) [inline Cringe/Cringe.csproj:16]: This can be removed

File: Cringe/Services/PpService.cs [hunk 3]
```diff
@@ -1,18 +1,35 @@
...
 namespace Cringe.Services
 {
     public class PpService
     {
         private readonly BeatmapService _beatmapService;
+
         private readonly ILogger<PpService> _logger;
 
+        //other game modes fuck you
+        private static readonly OsuRuleset _ruleset = new();
```
Comment (stanriders) [inline Cringe/Services/PpService.cs:31]: Rulesets are meant to be created in place, they're very lightweight. I don't think theres any reason to store it here

File: Cringe/Services/PpService.cs [hunk 4]
```diff
@@ -25,25 +42,87 @@ public async Task<double> CalculatePp(ScoreBase score)
             {
                 await using var stream =
                     new MemoryStream(await _beatmapService.GetBeatmapBytes(score.BeatmapId), false);
-                using var reader = new StreamReader(stream, true);
 
-                var map = Beatmap.Read(reader);
-                var diff = new DiffCalc().Calc(map, (Mods) score.Mods);
+                var beatmap = new StreamWorkingBeatmap(stream, score.BeatmapId);
+
+                var calculator = new OsuPerformanceCalculator();
```
Comment (stanriders) [inline Cringe/Services/PpService.cs:48]: `_ruleset.CreatePerformanceCalculator()`

File: Cringe/Services/PpService.cs [hunk 5]
```diff
@@ -25,25 +42,87 @@ public async Task<double> CalculatePp(ScoreBase score)
...
+
+                var mods = _ruleset.ConvertFromLegacyMods((LegacyMods) (int) score.Mods).Append(new OsuModClassic());
 
-                var pp = new PPv2(new PPv2Parameters(map, diff, score.Accuracy / 100, score.CountMiss, score.MaxCombo,
-                    (Mods) score.Mods)).Total;
+                var difficultyAttributesCalculator = new OsuDifficultyCalculator(_ruleset.RulesetInfo, beatmap);
```
Comment (stanriders) [inline Cringe/Services/PpService.cs:52]: `_ruleset.CreateDifficultyCalculator(beatmap)`

File: Cringe/Services/PpService.cs [hunk 6]
```diff
@@ -25,25 +42,87 @@ public async Task<double> CalculatePp(ScoreBase score)
...
+
+                var scoreModel = new ScoreInfo(beatmap.BeatmapInfo, _ruleset.RulesetInfo)
+                {
+                    Statistics = new Dictionary<HitResult, int>
+                    {
+                        [HitResult.Great] = score.Count300,
+                        [HitResult.Ok] = score.Count100,
+                        [HitResult.Meh] = score.Count50,
+                        [HitResult.Miss] = score.CountMiss,
+                        [HitResult.SliderTailHit] = difficultyAttributes.SliderCount,
```
Comment (stanriders) [inline Cringe/Services/PpService.cs:63]: `SliderTailHit` and `LargeTickMiss` (as well as some other hitresults you skipped) are lazer-only, no need to populate them

File: Cringe/Services/PpService.cs [hunk 7]
```diff
@@ -25,25 +42,87 @@ public async Task<double> CalculatePp(ScoreBase score)
...
+                        [HitResult.LargeTickMiss] = 0
+                    },
+                    MaxCombo = score.MaxCombo,
+                    Accuracy = score.Accuracy,
+                };
+
+                var performanceAttributes = await calculator.CalculateAsync(scoreModel, difficultyAttributes, CancellationToken.None);
```
Comment (stanriders) [inline Cringe/Services/PpService.cs:70]: There's no reason to run it async, replace with `Calculate`
image

File: Cringe/Services/PpService.cs [hunk 8]
```diff
@@ -25,25 +42,87 @@ public async Task<double> CalculatePp(ScoreBase score)
...
+        public StreamWorkingBeatmap(IBeatmap beatmap) : base(beatmap.BeatmapInfo, null)
+        {
+            _beatmap = beatmap;
+        }
+
+        private static IBeatmap readFromStream(Stream beatmapStream)
+        {
+            using var lineBufferedReader = new LineBufferedReader(beatmapStream);
+
+            return Decoder.GetDecoder<osu.Game.Beatmaps.Beatmap>(lineBufferedReader).Decode(lineBufferedReader);
```
Comment (stanriders) [inline Cringe/Services/PpService.cs:108]: You can do this in-place

```
using var lineBufferedReader = new LineBufferedReader(stream);
var beatmap = Decoder.GetDecoder<osu.Game.Beatmaps.Beatmap>(lineBufferedReader).Decode(lineBufferedReader);
var workingBeatmap = new FlatWorkingBeatmap(beatmap);
```

**Review by stanriders — CHANGES_REQUESTED** [review]

**Comment by stanriders** [comment]
Also add a way to recalculate already existing scores

File: Cringe.Web/Controllers/RootController.cs [hunk 9]
```diff
@@ -50,6 +53,31 @@ public async Task<IActionResult> DownloadBeatmap(string id, [FromQuery(Name = "u
...
+            _logger.LogWarning("Started PP recalculation");
+
+            var submittedScores = await databaseContext.RecentScores.ToListAsync();
+
+            foreach (var recentScore in submittedScores)
+            {
+                _logger.LogInformation("Recalculation: {ScoreId}", recentScore.Id);
+                var newPp = await ppService.CalculatePp(recentScore);
+                _logger.LogInformation("PP change {OldPp} -> {NewPp}", recentScore.Pp, newPp);
+                recentScore.Pp = newPp;
```
Comment (stanriders) [inline Cringe.Web/Controllers/RootController.cs:73]: `await databaseContext.Update(recentScore)`

Comment (JustRoxy) [reply Cringe.Web/Controllers/RootController.cs:76]: Unnecessary and lead to worse performance overall, see: https://learn.microsoft.com/en-us/ef/core/change-tracking/

**Review by stanriders — APPROVED** [review]

**Pull request merged**